syntax = "proto3";
package solhunter_zero;

message ActionExecuted {
  string action_json = 1;
  string result_json = 2;
}

message WeightsUpdated {
  map<string, double> weights = 1;
}

message RLWeights {
  map<string, double> weights = 1;
  map<string, double> risk = 2;
}

message RLCheckpoint {
  double time = 1;
  string path = 2;
}

message PortfolioUpdated {
  map<string, double> balances = 1;
}

message TokenInfo {
  double bids = 1;
  double asks = 2;
  double tx_rate = 3;
}

message TokenAgg {
  map<string, TokenInfo> dex = 1;
  double bids = 2;
  double asks = 3;
  double tx_rate = 4;
  int64 ts = 5;
}

message DepthUpdate {
  map<string, TokenAgg> entries = 1;
}

message DepthDiff {
  map<string, TokenAgg> entries = 1;
}

message DepthServiceStatus {
  string status = 1;
}

message Heartbeat {
  string service = 1;
}

message TradeLogged {
  string token = 1;
  string direction = 2;
  double amount = 3;
  double price = 4;
  string reason = 5;
  string context = 6;
  string emotion = 7;
  int64 simulation_id = 8;
  string uuid = 9;
  int64 trade_id = 10;

  // unix seconds (float64). Optional, omitted by older producers.
  double created_at = 11;
}

message RuntimeLog {
  string stage = 1;
  string detail = 2;
  double ts = 3;
  string level = 4;
  int32 actions = 5;
}

message RLMetrics {
  double loss = 1;
  double reward = 2;
}

message SystemMetrics {
  double cpu = 1;
  double memory = 2;
}

message PriceUpdate {
  string venue = 1;
  string token = 2;
  double price = 3;
}

message RouteRequest {
  string token = 1;
  double amount = 2;
  uint32 max_hops = 3;
}

message RouteResponse {
  repeated string path = 1;
  double profit = 2;
  double slippage = 3;
}

message ConfigUpdated {
  string config_json = 1;
}

message PendingSwap {
  string token = 1;
  string address = 2;
  double size = 3;
  double slippage = 4;
}

message RemoteSystemMetrics {
  double cpu = 1;
  double memory = 2;
}

message DoubleList {
  repeated double values = 1;
}

message RiskMetrics {
  double covariance = 1;
  double portfolio_cvar = 2;
  double portfolio_evar = 3;
  double correlation = 4;
  repeated DoubleList cov_matrix = 5;
  repeated DoubleList corr_matrix = 6;
}

message RiskUpdated {
  double multiplier = 1;
}

message SystemMetricsCombined {
  double cpu = 1;
  double memory = 2;
  double iter_ms = 3;
}

message TokenDiscovered {
  repeated string tokens = 1;
}

message MemorySyncRequest {
  int64 last_id = 1;
}

message MemorySyncResponse {
  repeated TradeLogged trades = 1;
  bytes index = 2;
}

message ActionProposal {
  string token = 1;
  string side = 2;
  double size = 3;
  double score = 4;
  string agent = 5;
  double price = 6;
}

message DecisionRationale {
  string agent = 1;
  repeated string agents = 2;
  double score = 3;
  double confidence = 4;
  double conviction_delta = 5;
  string snapshot_hash = 6;
}

message ActionDecision {
  string token = 1;
  string side = 2;
  double size = 3;
  double price = 4;
  DecisionRationale rationale = 5;
}

message DecisionSummary {
  string token = 1;
  uint32 count = 2;
  uint32 buys = 3;
  uint32 sells = 4;
}

message DecisionMetrics {
  double window_sec = 1;
  uint32 count = 2;
  double decision_rate = 3;
  uint32 buys = 4;
  uint32 sells = 5;
  double avg_size = 6;
  double ttf_decision = 7;
}

message DexLatencyUpdate {
  map<string, double> latencies = 1;
}

message RuntimeStageChanged {
  string stage = 1;
  bool ok = 2;
  string detail = 3;
}

message ScalarMetric {
  double value = 1;
}

message StartupComplete {
  double startup_duration_ms = 1;
}

message VirtualPnL {
  string order_id = 1;
  string mint = 2;
  string snapshot_hash = 3;
  double realized_usd = 4;
  double unrealized_usd = 5;
  double ts = 6;
}

message Event {
  string topic = 1;
  oneof kind {
    ActionExecuted action_executed = 2;
    WeightsUpdated weights_updated = 3;
    RLWeights rl_weights = 4;
    RLCheckpoint rl_checkpoint = 5;
    PortfolioUpdated portfolio_updated = 6;
    DepthUpdate depth_update = 7;
    DepthServiceStatus depth_service_status = 8;
    Heartbeat heartbeat = 9;
    TradeLogged trade_logged = 10;
    RLMetrics rl_metrics = 11;
    SystemMetrics system_metrics = 12;
    PriceUpdate price_update = 13;
    ConfigUpdated config_updated = 14;
    PendingSwap pending_swap = 15;
    RemoteSystemMetrics remote_system_metrics = 16;
    RiskMetrics risk_metrics = 17;
    RiskUpdated risk_updated = 18;
    SystemMetricsCombined system_metrics_combined = 19;
    TokenDiscovered token_discovered = 20;
    MemorySyncRequest memory_sync_request = 21;
    MemorySyncResponse memory_sync_response = 22;
    DepthDiff depth_diff = 23;

    RuntimeLog runtime_log = 24;
    ActionProposal action_proposal = 25;
    ActionDecision action_decision = 26;
    DecisionSummary decision_summary = 27;
    DecisionMetrics decision_metrics = 28;
    DexLatencyUpdate dex_latency_update = 29;
    RuntimeStageChanged runtime_stage_changed = 30;
    ScalarMetric scalar_metric = 31;
    StartupComplete startup_complete = 32;
    VirtualPnL virtual_pnl = 33;
  }
  string dedupe_key = 34;
}
